server/util/authutil/authutil.go

```diff
@@ -31,12 +44,41 @@ func AuthorizeGroupRole(u interfaces.UserInfo, groupID string, allowedRoles role
 	return nil
 }
 
+// AnonymousUserError returns an error indicating that the user is not
+// authenticated due to credentials being missing from the request.
+func AnonymousUserError(msg string) error {
+	info := &errdetails.ErrorInfo{Reason: missingCredentialsErrorReason}
+	status := gstatus.New(codes.Unauthenticated, msg)
+	if d, err := status.WithDetails(info); err != nil {
+		alert.UnexpectedEvent("failed_to_set_status_details", "Failed to set gRPC status details for AnonymousUserError")
+		return status.Err()
+	} else {
+		return d.Err()
+	}
+}
+
+// AnonymousUserError returns an error indicating that the user is not
+// authenticated due to credentials being missing from the request.
+func AnonymousUserErrorf(format string, args ...any) error {
+	return AnonymousUserError(fmt.Sprintf(format, args...))
+}
+
 // IsAnonymousUserError can be used to check whether an error returned by
 // functions which return the authenticated user (such as AuthenticatedUser or
 // AuthenticateSelectedGroupID) is due to an anonymous user accessing the
 // service. This is useful for allowing anonymous users to proceed, in cases
 // where anonymous usage is explicitly enabled in the app config, and we support
 // anonymous usage for the part of the service where this is used.
 func IsAnonymousUserError(err error) bool {
-	return status.IsUnauthenticatedError(err) || status.IsPermissionDeniedError(err) || status.IsUnimplementedError(err)
+	for _, detail := range gstatus.Convert(err).Proto().GetDetails() {
+		info := &errdetails.ErrorInfo{}
+		if err := detail.UnmarshalTo(info); err != nil {
+			// not an ErrorInfo detail; ignore.
+			continue
+		}
+		if info.GetReason() == missingCredentialsErrorReason {
+			return true
+		}
+	}
+	return false
 }
```

Comment:
note: I changed this implementation to check for an explicit error status created by `AnonymousUserError`, so that we can have stronger guarantees about anonymous user errors vs. other types of errors that we happen to be setting to "Unauthenticated." I felt this is important since now we are leaning more heavily on AnonymousUserError in the cache prefix code, but let me know if I am overthinking it and if it seems fine to just use the old check instead.